LUCENE-9935: Enable bulk merge for stored fields with index sort

dnhatn opened the pull request:
This commit enables bulk-merges (i.e., raw chunk copying) for stored fields when index sort is enabled. I benchmarked this change with Elasticsearch using Rally and found that this reduced the merge time by 20% (from  14.9 min ->  11.9 min).
<details>
<summary>Benchmark details</summary>

|                            Metric |     Task |  Baseline |  Contender |   Diff |  Unit |
|--------------------------------------------------------------:|-------------:|------------:|------------:|---------:|-------:|
|          Cumulative indexing time of primary shards |       |   27.8208 |   26.4527 | -1.36817 |  min |
|       Min cumulative indexing time across primary shard |       |   4.21395 |   3.90437 | -0.30958 |  min |
|     Median cumulative indexing time across primary shard |       |   4.68754 |   4.43102 | -0.25652 |  min |
|       Max cumulative indexing time across primary shard |       |   5.00328 |   5.11322 | 0.10993 |  min |
|      Cumulative indexing throttle time of primary shards |       |   47.5573 |   43.7704 | -3.7869 |  min |
|  Min cumulative indexing throttle time across primary shard |       |   5.82492 |   6.24037 | 0.41545 |  min |
| Median cumulative indexing throttle time across primary shard |       |   7.86763 |   6.94697 | -0.92066 |  min |
|  Max cumulative indexing throttle time across primary shard |       |   9.2532 |   8.78763 | -0.46557 |  min |
|            Cumulative merge time of primary shards |       |   14.9712 |   11.9894 | -2.9818 |  min |
|           Cumulative merge count of primary shards |       |     52 |     49 |    -3 |    |
|        Min cumulative merge time across primary shard |       |   1.54965 |   1.2836 | -0.26605 |  min |
|       Median cumulative merge time across primary shard |       |   2.7143 |   1.99087 | -0.72343 |  min |
|        Max cumulative merge time across primary shard |       |   2.96187 |    3.013 | 0.05113 |  min |
|       Cumulative merge throttle time of primary shards |       |   0.0308 |  0.118767 | 0.08797 |  min |
|    Min cumulative merge throttle time across primary shard |       | 0.00116667 | 0.00716667 |  0.006 |  min |
|  Median cumulative merge throttle time across primary shard |       | 0.00191667 |  0.0116167 |  0.0097 |  min |
|    Max cumulative merge throttle time across primary shard |       |  0.0175333 |   0.0651 | 0.04757 |  min |
|           Cumulative refresh time of primary shards |       |   15.7809 |   15.123 | -0.65797 |  min |
|          Cumulative refresh count of primary shards |       |    1786 |    1843 |    57 |    |
|       Min cumulative refresh time across primary shard |       |   2.15032 |   2.12483 | -0.02548 |  min |
|      Median cumulative refresh time across primary shard |       |   2.6072 |   2.55063 | -0.05657 |  min |
|       Max cumulative refresh time across primary shard |       |   3.07643 |   2.76125 | -0.31518 |  min |
|            Cumulative flush time of primary shards |       |      0 |      0 |    0 |  min |
|           Cumulative flush count of primary shards |       |      0 |      0 |    0 |    |
|        Min cumulative flush time across primary shard |       |      0 |      0 |    0 |  min |
|       Median cumulative flush time across primary shard |       |      0 |      0 |    0 |  min |
|        Max cumulative flush time across primary shard |       |      0 |      0 |    0 |  min |
|                    Total Young Gen GC time |       |   37.456 |   38.081 |  0.625 |   s |
|                   Total Young Gen GC count |       |    1407 |    1379 |   -28 |    |
|                     Total Old Gen GC time |       |      0 |      0 |    0 |   s |
|                    Total Old Gen GC count |       |      0 |      0 |    0 |    |
|                          Store size |       |   2.77534 |   2.89203 | 0.11669 |   GB |
|                         Translog size |       | 3.07336e-07 | 3.07336e-07 |    0 |   GB |
|                    Heap used for segments |       |  0.898968 |  0.821392 | -0.07758 |   MB |
|                   Heap used for doc values |       |  0.0717201 |  0.0275154 | -0.0442 |   MB |
|                      Heap used for terms |       |  0.676544 |  0.648956 | -0.02759 |   MB |
|                      Heap used for norms |       |   0.09198 |  0.0881958 | -0.00378 |   MB |
|                     Heap used for points |       |      0 |      0 |    0 |   MB |
|                  Heap used for stored fields |       |  0.0587234 |  0.0567245 |  -0.002 |   MB |
|                         Segment count |       |     119 |     115 |    -4 |    |
|                        Min Throughput | index-append |   11692.2 |   12325.1 | 632.907 | docs/s |
|                        Mean Throughput | index-append |   12580.3 |   12918.5 | 338.218 | docs/s |
|                       Median Throughput | index-append |   12602.7 |   12950.9 | 348.173 | docs/s |
|                        Max Throughput | index-append |   13131.5 |   13787.7 | 656.203 | docs/s |
|                    50th percentile latency | index-append |   2917.01 |   2920.49 | 3.48737 |   ms |
|                    90th percentile latency | index-append |   4142.28 |   3969.75 | -172.527 |   ms |
|                    99th percentile latency | index-append |   15025.7 |   12156.8 | -2868.93 |   ms |
|                   99.9th percentile latency | index-append |   17450.6 |   12870.4 | -4580.23 |   ms |
|                   100th percentile latency | index-append |   18089.8 |   13110.4 | -4979.42 |   ms |
|                 50th percentile service time | index-append |   2917.01 |   2920.49 | 3.48737 |   ms |
|                 90th percentile service time | index-append |   4142.28 |   3969.75 | -172.527 |   ms |
|                 99th percentile service time | index-append |   15025.7 |   12156.8 | -2868.93 |   ms |
|                99.9th percentile service time | index-append |   17450.6 |   12870.4 | -4580.23 |   ms |
|                 100th percentile service time | index-append |   18089.8 |   13110.4 | -4979.42 |   ms |

</details>

I also tried to enable bulk merge with deletions but the result wasn't great. I will explore an idea that iterates the start pointers of chunks sequentially to avoid the cost of searching.

dnhatn committed:
Enable bulk merge for stored fields with index sort

dnhatn commented:
@rmuir It would be great if you can take a look at this PR. Thank you!

dnhatn committed:
spotless

rmuir commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java:
```diff
@@ -76,7 +75,8 @@
   static final int VERSION_START = 1;
-  static final int VERSION_CURRENT = VERSION_START;
+  static final int VERSION_TRACK_DIRTY_CHUNK = 2;
+  static final int VERSION_CURRENT = VERSION_TRACK_DIRTY_CHUNK;
```
because lucene 9.0 is not yet released, we shouldn't bump any internal versions as we don't need backwards compatibility with them here. can remove this part and other conditionals related to it elsewhere.

dnhatn replied:
Good catch! I removed it in https://github.com/apache/lucene/pull/134/commits/fc4e8c381388d6b6a649fa9e6757d92dd620fa44.

rmuir commented:
Change looks great, a few questions:
* Can we add tests to `BaseStoredFieldsFormatTestCase` that call `setIndexSort`, exercise merging, and validate correctness (e.g. check that stored fields values are still correct)? I am worried that existing test coverage for this change is not great.
* Can we also fix merge logic of `Lucene90TermVectorsFormat`/`BaseTermVectorsFormatTestCase`? Generally, we try to keep this logic consistent across stored fields and term vectors as much as possible, and this improvement should apply there as well?

dnhatn committed:
remove bwc

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java:
```diff
@@ -472,7 +479,7 @@ public void finish(FieldInfos fis, int numDocs) throws IOException {
     if (numBufferedDocs > 0) {
       numDirtyChunks++; // incomplete: we had to force this flush
       numDirtyDocs += numBufferedDocs;
```
Maybe we could move the accounting of `numDirtyChunks` and `numDirtyDocs` to `flush` now to make things simpler?

dnhatn replied:
++, I made this change in https://github.com/apache/lucene/pull/134/commits/b3cad2ecb01b6914433be442fa7e40f5a8a3ab88

jpountz reviewed (commented):
This looks very good. I like that the same merge logic is taking care of both the sorted and unsorted case.

dnhatn committed:
account dirty chunks in flush

dnhatn committed:
add indexSort test to BaseStoredFieldsFormatTestCase

dnhatn committed:
add CHANGES entry

dnhatn committed:
Merge branch 'main' into bulk-merge-storedfields

dnhatn commented:
@rmuir @jpountz Thanks for reviewing.

> Can we add tests to BaseStoredFieldsFormatTestCase that call setIndexSort, exercise merging, and validate correctness (e.g. check that stored fields values are still correct)? I am worried that existing test coverage for this change is not great.

I added a random test in https://github.com/apache/lucene/pull/134/commits/be7c71c320be86f5e485a6d0864b18954f5ef25f.

> Can we also fix merge logic of Lucene90TermVectorsFormat/BaseTermVectorsFormatTestCase? Generally, we try to keep this logic consistent across stored fields and term vectors as much as possible, and this improvement should apply there as well?

+1. Are you okay if I do will it in a follow-up right after this PR?

Would you please take another look?

rmuir commented:
> +1. Are you okay if I do will it in a follow-up right after this PR?

That's good, I just don't want to forget it. Main motivation is to only maintain one version of the algorithm and not to let the term-vectors one fall behind.

> Would you please take another look?

I'll take another pass and try to run some tests etc on my side as soon as I can.

rmuir committed:
LUCENE-9935: add help doc for evaluating coverage of particular test

This is a common use-case to confirm that your test is really exercising
the code that you think it is!

rmuir committed:
LUCENE-9935: ensure test exercises both MergeStrategy.BULK and MergeStrategy.DOC

Previously, only 50% of test runs would check the DOC case. The problem
is that sometimes the test would never issue any deletions at all.

Improve the test to always test both cases. Deletions are issued rarely
enough that we still always exercise the bulk case if we just turn them
on always.

rmuir commented:
I added a small tweak to the test so that it always tests both the bulk and non-bulk cases. I just ran the test a few times before/after and inspected its test coverage each run like this:

```console
$ ./gradlew -p lucene/core test --tests org.apache.lucene.codecs.lucene90.TestLucene90StoredFieldsFormat.testRandomStoredFieldsWithIndexSort coverage
$ open ~/workspace/lucene/lucene/core/build/reports/jacoco/test/html/index.html
```

in order to also exercise the `MergeStrategy.VISITOR` case, simplest approach might be to use `addIndexes` at the end. Or just add a separate unit test for that case. I am looking into it...

rmuir committed:
LUCENE-9935: always test MergeStrategy.VISITOR with index-time sort

There are 3 merge algorithms:
* BULK (fully optimized: no deletions, copy without recompressing)
* DOC (semi-optimized: deletions or dirty, but same codec version)
* VISITOR (generic: foreign codec or older index version)

Enforce that the generic algorithm gets tested too, using addIndexes.
This is important for codecs to implement or backwards compatibility
can't work.

rmuir reviewed (approved):
Great work here!

I ran a basic worst-case performance check with geonames (the standalone Indexer.java from https://issues.apache.org/jira/browse/LUCENE-9827) which uses BEST_COMPRESSION and flushes on every document and confirmed there's no slowdown for the default case of unsorted index. So I just want to echo @jpountz comment that it is really great there is no special casing for sorted/unsorted and the algorithm just becomes more generic.

I tested out the change and inspected coverage: I pushed a few very minor changes to the test to increase the coverage so that all 3 merge algorithms are always tested for the sorted index case. It is just a bit of paranoia due to historical bugs in bulk merge being really awful on the users: we want to be cautious. At the same time, it is really necessary for performance to not constantly recompress data over and over again.

dnhatn commented:
@rmuir @jpountz Thank you so much for your reviews and help in this PR! I really appreciate that.
